src/modules/Game.class.js

```diff
@@ -50,19 +118,221 @@ class Game {
+
+    this.initialState = returnToInitialArray;
+    this.score = 0;
+    this.state = 'playing';
+
+    return this.initialState;
+  }
 
   /**
    * Resets the game.
    */
-  restart() {}
+  restart() {
+    this.initialState = [
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+      [0, 0, 0, 0],
+    ];
+    this.score = 0;
+    this.state = 'idle';
+
+    return this.initialState;
+  }
 
   // Add your own methods here
 }
 
+function addNewCellRandomCell(initialState, qntOfRandomCellsToAdd) {
+  const flatGameInARow = initialState.flat();
+  const availableIndices = flatGameInARow
+    .map((value, i) => (value === 0 ? i : null))
+    .filter((value) => value !== null);
+
+  if (availableIndices.length === 0) {
+    return initialState;
+  }
+
+  for (let i = 0; i < qntOfRandomCellsToAdd; i++) {
+    if (availableIndices.length === 0) {
+      break;
+    }
+
+    const randomIndex = Math.floor(Math.random() * availableIndices.length);
+    const indexToAdd = availableIndices.splice(randomIndex, 1)[0];
+    const randomValue = Math.random() < 0.9 ? 2 : 4;
+
+    flatGameInARow[indexToAdd] = randomValue;
+  }
+
+  const returnToInitialArray = [];
+
+  while (flatGameInARow.length) {
+    returnToInitialArray.push(flatGameInARow.splice(0, 4));
+  }
+
+  return returnToInitialArray;
+}
+
+function mergeToTheLeft(result) {
+  for (let i = 0; i < result.length - 1; i++) {
+    if (result[i] === result[i + 1] && result[i] !== 0) {
+      result[i] *= 2;
+      result[i + 1] = 0;
+    }
+  }
+
+  const filtered = result.filter((num) => num !== 0);
+
+  while (filtered.length < result.length) {
+    filtered.push(0);
+  }
+
+  return filtered;
+}
+
+function mergeToTheRight(result) {
+  let nonZeroValues = result.filter((value) => value !== 0);
+
+  for (let i = nonZeroValues.length - 1; i > 0; i--) {
+    if (nonZeroValues[i] === nonZeroValues[i - 1]) {
+      nonZeroValues[i] *= 2;
+      nonZeroValues[i - 1] = 0;
+    }
+  }
+  // Remove zeros again after merging
+  nonZeroValues = nonZeroValues.filter((value) => value !== 0);
+
+  // Add zeros to the left to maintain the row length
+  while (nonZeroValues.length < result.length) {
+    nonZeroValues.unshift(0);
+  }
+
+  return nonZeroValues;
+}
+
+function mergeDown(initialState) {
+  const result = initialState.map((col) => [...col]);
+
+  for (let col = 0; col < result[0].length; col++) {
+    const column = result.map((row) => row[col]);
+
+    for (let i = column.length - 1; i > 0; i--) {
+      if (column[i] === column[i - 1] && column[i] !== 0) {
+        column[i] *= 2;
+        column[i - 1] = 0;
+      }
+    }
+
+    const filtered = column.filter((num) => num !== 0);
+
+    while (filtered.length < column.length) {
+      filtered.unshift(0);
+    }
+
+    for (let i = 0; i < result.length; i++) {
+      result[i][col] = filtered[i];
+    }
+  }
+
+  return result;
+}
+
+function mergeUp(initialState) {
+  const result = initialState.map((col) => [...col]);
```

Comment:
The `mergeUp` function creates a shallow copy of each row, which is unnecessary since you are working with columns. Consider revising this logic to correctly handle column merging.